### DISCUSSION TITLE
File content mime type / extension

### DISCUSSION THREAD

**Discussion by MohammadZarifiyan** [laravel/framework#53736] [Ideas]
Dynamically get file content extension and mime type.
Should be place in `\Illuminate\Filesystem\Filesystem`:
```
use Illuminate\Http\Testing\MimeType;

public function contentExtension($content): string
{
    $mime_type = $this->contentMimeType($content);

    return MimeType::search($mime_type);
}

public function contentMimeType($content): string
{
    if (is_resource($content)) {
        $content = stream_get_contents($content);
    }

    $finfo = finfo_open(FILEINFO_MIME_TYPE);
    $result = finfo_buffer($finfo, $content);
    finfo_close($finfo);

    return $result;
}
```

**Comment by macropay-solutions** [comment]
See
\Illuminate\Support\Facades\File::extension(..)
\Symfony\Component\HttpFoundation\File\UploadedFile::getClientOriginalExtension
\Symfony\Component\HttpFoundation\File\File::guessExtension

**Reply by MohammadZarifiyan** [reply]
> \Illuminate\Support\Facades\File::extension(..)
> \Symfony\Component\HttpFoundation\File\UploadedFile::getClientOriginalExtension
> \Symfony\Component\HttpFoundation\File\File::guessExtension

All of these methods work when the file is stored on the server. The one I mentioned is used when the contents of a file are stored in a variable.

**Reply by macropay-solutions** [reply]
Ok.
If it is not a resource then, (if you agree):
```php

use Illuminate\Http\Testing\MimeType;

public function contentExtension(string $content): ?string
{
    return MimeType::search($this->contentMimeType($content));
}

public function contentMimeType(string $content): string
{
    $finfo = finfo_open(FILEINFO_MIME_TYPE);
    $result = finfo_buffer($finfo, $content);
    finfo_close($finfo);

    return (string)$result;
}
```
or
```php

use Illuminate\Http\Testing\MimeType;

public function contentExtension(string $content): ?string
{
    return MimeType::search($this->contentMimeType($content));
}

/**
 * @param resource|string $content 
 */
public function contentMimeType(mixed $content): string
{
    if (is_resource($content)) {
        return (string)mime_content_type($content);
    }

    if (is_string($content)) {
        $content = finfo_open(FILEINFO_MIME_TYPE);
        $result = finfo_buffer($content, $content);
        finfo_close($content);

        return (string)$result;
    }

    return '';
}
```
you would also need 
`      "ext-fileinfo": "*"`
 in your composer.

**Reply by MohammadZarifiyan** [reply]
Ok? i'm telling that Laravel team can add my code to Laravel framework core. I'm not seeking for help!

**Reply by macropay-solutions** [reply]
Your code returns string and you wrote : Closure ...

**Reply by MohammadZarifiyan** [reply]
> Your code returns string and you wrote : Closure ...

Oh thank you, I edited it.

**Reply by shaedrich** [reply]
> Ok? i'm telling that Laravel team can add my code to Laravel framework core. I'm not seeking for help!

For that, you would need to [open a PR](https://github.com/laravel/framework/pulls), but don't get your hopes up too high
